### PR TITLE
Deep learning

### PR DESCRIPTION
**Pull request by Ulfgard**
LArge Rewrite of Shark Models to allow bette ruse in a deep learning context

### CODE REVIEW

File: src/Models/CMAC.cpp [hunk 1]
```diff
@@ -66,22 +66,23 @@ std::vector<std::size_t> CMACMap::getIndizes(blas::matrix_row< const RealMatrix>
...
 		bounds(dim, 0) = lower;
 		bounds(dim, 1) = upper;
 	}
 	setStructure(inputs, outputs, numberOfTilings, numberOfTiles,bounds,randomTiles);
 }
-void CMACMap::setStructure(std::size_t inputs, std::size_t outputs, std::size_t numberOfTilings, std::size_t numberOfTiles, RealMatrix const& bounds, bool randomTiles){
-	m_inputSize  = inputs;
-	m_outputSize = outputs;
+void CMACMap::setStructure(Shape const& inputs, Shape const& outputs, std::size_t numberOfTilings, std::size_t numberOfTiles, RealMatrix const& bounds, bool randomTiles){
+	m_inputSize  = inputs.numElements();
```
Comment (TGlas) [inline src/Models/CMAC.cpp]: m_inputSize and m_inputShape are highly redundant, in particular since the shape already keeps the size as a redundant data member. Can we remove m_inputSize?

File: Test/Models/ConcatenatedModel.cpp [hunk 2]
```diff
@@ -43,49 +44,58 @@ BOOST_AUTO_TEST_CASE( CONCATENATED_MODEL_Value )
...
+	LinearModel<> net3;
+	net1.setStructure(3,5,1);
+	net3.setStructure(5,2,1);
+	ConcatenatedModel<RealVector> model  = net1 >> net2 >> net3;
 
 	//test1: all activated
 	{
 		//parameters
-		size_t modelParameters = net1.numberOfParameters()+net2.numberOfParameters();
+		size_t modelParameters = net1.numberOfParameters()+net3.numberOfParameters();
```
Comment (TGlas) [inline Test/Models/ConcatenatedModel.cpp]: If this case is "all activated" as the comment states, then we should also add the number of parameters of net2.

File: Test/Models/ConcatenatedModel.cpp [hunk 3]
```diff
@@ -219,108 +313,28 @@ BOOST_AUTO_TEST_CASE( CONCATENATED_MODEL_SERIALIZE )
...
+	BOOST_REQUIRE_EQUAL(net1.inputShape(),netTest1.inputShape());
+	BOOST_REQUIRE_EQUAL(net1.outputShape(),netTest1.outputShape());
+	BOOST_REQUIRE_EQUAL(net2.inputShape(),netTest2.inputShape());
+	BOOST_REQUIRE_EQUAL(net2.outputShape(),netTest2.outputShape());
+	BOOST_REQUIRE_EQUAL(net3.inputShape(),netTest3.inputShape());
+	BOOST_REQUIRE_EQUAL(net3.outputShape(),netTest3.outputShape());
 	for (size_t i=0; i<1000; i++)
 	{
 		RealVector output = modelDeserialized(dataset.element(i).input);
 		BOOST_CHECK_SMALL(norm_2(output -dataset.element(i).label),1.e-50);
```
Comment (TGlas) [inline Test/Models/ConcatenatedModel.cpp]: "1.e-50" - for labels? With double precision, for all practical purposes, that's zero - right? (Unless we have really miniscule regression labels.)

File: doc/sphinx_pages/rest_sources/tutorials/concepts/misc/versatile_classification.tut [hunk 4]
```diff
@@ -1,190 +0,0 @@
-
```
Comment (TGlas) [inline doc/sphinx_pages/rest_sources/tutorials/concepts/misc/versatile_classification.tut]: Is this removed because it is not needed/wanted any more, or becaue it needs an update?

File: examples/Supervised/VersatileClassificationTutorial-LDA.tpp [hunk 5]
```diff
@@ -1,54 +0,0 @@
-
```
Comment (TGlas) [inline examples/Supervised/VersatileClassificationTutorial-LDA.tpp]: Why is this removed? LDA is a very useful method. It is particularly nice for ML beginners, because it does not have any parameters.

File: doc/sphinx_pages/rest_sources/tutorials/concepts/misc/versatile_classification.tut (hunk 4 above)
Comment (Ulfgard) [reply doc/sphinx_pages/rest_sources/tutorials/concepts/misc/versatile_classification.tut]: i removed the versatile_* tutorials because they would need an update and a "look in how many different APIs you can classify" tutorial is maybe not very helpful. The idea in parts of the rework of the last months  was to make the APIs so consistent that you only need the distinction "do i train a neural network or can i use a premade trainer?" which made this obsolet

File: examples/Supervised/VersatileClassificationTutorial-LDA.tpp (hunk 5 above)
Comment (Ulfgard) [reply examples/Supervised/VersatileClassificationTutorial-LDA.tpp]: LDA is covered in the getting started tutorial.

File: include/shark/Core/Shape.h [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+public:
+	Shape(): m_numElements(1){}
+	Shape(std::size_t size): m_dims(1,size), m_numElements(size){}
+	Shape(std::initializer_list<std::size_t> dims): m_dims(dims){
+		m_numElements = 1;
+		for(auto dim: m_dims){
+			m_numElements *= dim;
+		}
+	}
+	std::size_t numDims()const{
```
[the remark this answers is not in the document]
Comment (Ulfgard) [reply include/shark/Core/Shape.h]: i opted against this maybe because "size" in this context is quite overloaded (e.g. opencv would use size() in a matrix to return the shape). But yeah, maybe i should change the interface.

File: include/shark/Models/AbstractModel.h [hunk 7]
```diff
@@ -98,32 +104,23 @@ class AbstractModel : public IParameterizable, public INameable, public ISeriali
...
-		HAS_SECOND_INPUT_DERIVATIVE     = 8,
-		IS_SEQUENTIAL = 16
 	};
 	SHARK_FEATURE_INTERFACE;
 
 	/// \brief Returns true when the first parameter derivative is implemented.
 	bool hasFirstParameterDerivative()const{
 		return m_features & HAS_FIRST_PARAMETER_DERIVATIVE;
 	}
-	/// \brief Returns true when the second parameter derivative is implemented.
```
[the remark this answers is not in the document]
Comment (Ulfgard) [reply include/shark/Models/AbstractModel.h]: the problem is that we did not have an (consistent) interface for second derivatives in the first place. Returning a vector of matrices(one for every model output) is not very efficient and most of the time we would only look at linear combinations of the form sum_i w_i H_i v (w_i scalar and v a vector). Without an interface we do not need the flag.

File: include/shark/Models/ConvolutionalModel.h [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+	
+	///\brief Sets the structure by setting the dimensionalities of image and filters.
+	///
+	/// \arg imageShape Shape of the image imHeight x imWidth x channel
+	/// \arg filterShape Shape of the filter matrix numFilters x fiHeight x fiWidth x channel
+	Conv2DModel(
+		Shape const& imageShape, Shape const& filterShape
+	){
+		base_type::m_features |= base_type::HAS_FIRST_PARAMETER_DERIVATIVE;
+		base_type::m_features |= base_type::HAS_FIRST_INPUT_DERIVATIVE;
```
[the remark this answers is not in the document]
Comment (Ulfgard) [reply include/shark/Models/ConvolutionalModel.h]: i think having sparse convolutions does not make sense i will remove the first check

File: include/shark/Models/ConvolutionalModel.h [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+	std::string name() const
+	{ return "Conv2DModel"; }
+	
+	///\brief Returns the expected shape of the input
+	Shape inputShape() const{
+		return {m_imageHeight, m_imageWidth, m_numChannels};
+	}
+	///\brief Returns the shape of the output
+	Shape outputShape() const{
+		return {m_imageHeight, m_imageWidth, m_numFilters};
```
[the remark this answers is not in the document]
Comment (Ulfgard) [reply include/shark/Models/ConvolutionalModel.h]: tue, i changed this. I think by now zero-padding is more canonical.

**Comment by Ulfgard** [comment]
thank you very much! will fix :)

**Comment by TGlas** [comment]
>> +	Conv2DModel(){
> +		base_type::m_features |= base_type::HAS_FIRST_PARAMETER_DERIVATIVE;
> +		if(std::is_same<typename VectorType::storage_type::storage_tag, blas::dense_tag>::value){
> +			base_type::m_features |= base_type::HAS_FIRST_INPUT_DERIVATIVE;
> +		}
> +	}
> +	
> +	///\brief Sets the structure by setting the dimensionalities of image and filters.
> +	///
> +	/// \arg imageShape Shape of the image imHeight x imWidth x channel
> +	/// \arg filterShape Shape of the filter matrix numFilters x fiHeight x fiWidth x channel
> +	Conv2DModel(
> +		Shape const& imageShape, Shape const& filterShape
> +	){
> +		base_type::m_features |= base_type::HAS_FIRST_PARAMETER_DERIVATIVE;
> +		base_type::m_features |= base_type::HAS_FIRST_INPUT_DERIVATIVE;
> 
> i think having sparse convolutions does not make sense i will remove the first check

Sparse filters don't, sparse inputs may make sense? Not sure though.

**Comment by TGlas** [comment]
> Ulfgard commented on this pull request.
> 
> 
> 
>> +	){
> +		base_type::m_features |= base_type::HAS_FIRST_PARAMETER_DERIVATIVE;
> +		base_type::m_features |= base_type::HAS_FIRST_INPUT_DERIVATIVE;
> +		setStructure(imageShape, filterShape);
> +	}
> +
> +	std::string name() const
> +	{ return "Conv2DModel"; }
> +	
> +	///\brief Returns the expected shape of the input
> +	Shape inputShape() const{
> +		return {m_imageHeight, m_imageWidth, m_numChannels};
> +	}
> +	///\brief Returns the shape of the output
> +	Shape outputShape() const{
> +		return {m_imageHeight, m_imageWidth, m_numFilters};
> 
> tue, i changed this. I think by now zero-padding is more canonical.

How much effort would it be to keep the padding flexible?
If I understand correctly then we need padding anyway for the
reverse convolution in backprop. So how about a "padding"
parameter specifying the about of padding, which could default
to half of the kernel size?

**Comment by Ulfgard** [comment]
It is not very complicated. but instead of specifying the padding we should rather define the type (pad to same size with 0 or no padding).

________________________________

File: include/shark/Models/DropoutLayer.h [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+		SIZE_CHECK(coefficients.size1() == patterns.size1());
+		SIZE_CHECK(coefficients.size2() == patterns.size2());
+
+		MatrixType const& mask = state.toState<InternalState>().mask;
+		derivative.resize(coefficients.size1(),coefficients.size2());
+		noalias(derivative) = coefficients * mask;
+	}
+
+	/// From ISerializable
+	void read(InArchive& archive){ archive >> m_dropoutProbability;}
```
[the remark this answers is not in the document]
Comment (Ulfgard) [reply include/shark/Models/DropoutLayer.h]: one can not construct it without a shape. so yes

File: include/shark/Models/ConvolutionalModel.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (Ulfgard) [reply include/shark/Models/ConvolutionalModel.h]: there is no implementation of sparse convolutions so i just added a compile time check that the vector is dense.

File: include/shark/Models/ConvolutionalModel.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (Ulfgard) [reply include/shark/Models/ConvolutionalModel.h]: i added another parameter for either "Convolution::Valid" or "Convolution::ZeroPad"

File: include/shark/Core/Shape.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (Ulfgard) [reply include/shark/Core/Shape.h]: i have now changed the interface to use size() and op[]

**Comment by Ulfgard** [comment]
I am pushing and emrging soon. I also removed the ConvexCombination for now, because it is not really relevant.

so next step is merging & implementing pooling.
